### PR TITLE
deprecating nstemplateset fields 

### PR DESCRIPTION
**Pull request by tinakurian**
removing revision, type and tiername from nstemplate set. 

Jira: https://issues.redhat.com/browse/CRT-503
related PRs:
api - https://github.com/codeready-toolchain/api/pull/143
member - https://github.com/codeready-toolchain/member-operator/pull/164

### CODE REVIEW

**Comment by MatousJobanek** [comment]
As I wrote in the API PR - the `TierName` field should stay. https://github.com/codeready-toolchain/api/pull/143#discussion_r432007565
In addition, removing from the CRDs is not sufficient - you should drop also the usage from the go code.

**Comment by tinakurian** [comment]
> As I wrote in the API PR - the `TierName` field should stay. [codeready-toolchain/api#143 (comment)](https://github.com/codeready-toolchain/api/pull/143#discussion_r432007565)
> In addition, removing from the CRDs is not sufficient - you should drop also the usage from the go code.

fixed here: a605f141110c8ca675959ec188112f620eb8dd6d

**Comment by MatousJobanek** [comment]
> fixed here: a605f14

great but don't forget to drop also the usage of the removed fields from the go code...

**Comment by tinakurian** [comment]
> > fixed here: [a605f14](https://github.com/codeready-toolchain/host-operator/commit/a605f141110c8ca675959ec188112f620eb8dd6d)
> 
> great but don't forget to drop also the usage of the removed fields from the go code...

yeah, didn't want to do that all in the same PR for a couple of reasons: 
1. this requires changes in both host and member operator and since we can only pair one at a time, i figured it would make sense to first do the api changes and then raise the PRs one at a time for member and then for host
2. i thought it would also reduce the amount of code that needed to be reviewed (smaller PRs)

**Comment by alexeykazakov** [comment]
> yeah, didn't want to do that all in the same PR for a couple of reasons

You have to drop the removed field usage in the code too. The code will fail in runtime if you try to use fields not specified in CRDs. The e2e tests should fail because of that too.

**Comment by MatousJobanek** [comment]
> this requires changes in both host and member operator and since we can only pair one at a time, i figured it would make sense to first do the api changes and then raise the PRs one at a time for member and then for host

The member-operator should not contain any usage of these fields, so dropping the usage only in host-operator should be enough. You need to open PRs for both operators for sure as well as for the e2e repo - the pairing should be done only with host-operator PR though.

**Comment by tinakurian** [comment]
> > this requires changes in both host and member operator and since we can only pair one at a time, i figured it would make sense to first do the api changes and then raise the PRs one at a time for member and then for host
> 
> The member-operator should not contain any usage of these fields, so dropping the usage only in host-operator should be enough. You need to open PRs for both operators for sure as well as for the e2e repo - the pairing should be done only with host-operator PR though.

I'm looking at the changes i have made and you're right! the changes in member were pertaining to the tierName but since we're keeping that i guess i can push my code changes here for host. will update that! thanks @MatousJobanek!

File: pkg/controller/masteruserrecord/masteruserrecord_controller_test.go [hunk 1]
```diff
@@ -384,7 +384,7 @@ func TestModifyUserAccounts(t *testing.T) {
 	userAccount2 := uatest.NewUserAccountFromMur(mur)
 	userAccount3 := uatest.NewUserAccountFromMur(mur)
 
-	murtest.ModifyUaInMur(mur, test.MemberClusterName, murtest.NsLimit("advanced"), murtest.TierName("admin"), murtest.Namespace("ide", "54321"))
+	murtest.ModifyUaInMur(mur, test.MemberClusterName, murtest.NsLimit("advanced"), murtest.TierName("admin"), murtest.Namespace("basic-dev-123abc", "admin-ide-54321"))
```
Comment (MatousJobanek) [inline pkg/controller/masteruserrecord/masteruserrecord_controller_test.go:387]: do you know that you are replacing different tier and different namespace type?

File: pkg/controller/masteruserrecord/sync_test.go [hunk 2]
```diff
@@ -86,16 +86,9 @@ func setupSynchronizerItems() (toolchainv1alpha1.MasterUserRecord, toolchainv1al
...
-			},
-			Namespaces: []toolchainv1alpha1.NSTemplateSetNamespace{
-				{
-					Revision: "123456a",
-					Type:     "code",
-				},
-			},
+			TierName:         "basic",
+			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{},
+			Namespaces:       []toolchainv1alpha1.NSTemplateSetNamespace{},
```
Comment (MatousJobanek) [inline pkg/controller/masteruserrecord/sync_test.go:91]: This should contain a namespace with `TemplateRef` if I'm not mistaken

File: pkg/controller/masteruserrecord/sync_test.go [hunk 3]
```diff
@@ -131,7 +124,7 @@ func TestSynchronizeSpec(t *testing.T) {
 	userAccount := uatest.NewUserAccountFromMur(mur)
 
 	murtest.ModifyUaInMur(mur, test.MemberClusterName, murtest.NsLimit("advanced"),
-		murtest.TierName("admin"), murtest.Namespace("ide", "54321"))
+		murtest.TierName("admin"), murtest.Namespace("basic-dev-123abc", "admin-ide-54321"))
```
Comment (MatousJobanek) [inline pkg/controller/masteruserrecord/sync_test.go:127]: same here, you are replacing with different tier and with a different namespace type

File: pkg/controller/usersignup/masteruserrecord_test.go [hunk 4]
```diff
@@ -104,56 +104,6 @@ func TestMigrateMurIfNecessary(t *testing.T) {
...
-		mur.Spec.UserAccounts[0].Spec.NSTemplateSet.Namespaces[0].TemplateRef = "advanced-cicd-123abc1"
-		providedMur := mur.DeepCopy()
-
-		// when
-		changed, changedMur := migrateMurIfNecessary(mur, nSTemplateTier)
-
-		// then
-		assert.False(t, changed)
-		assert.Equal(t, *providedMur, *changedMur)
-	})
```
Comment (MatousJobanek) [inline pkg/controller/usersignup/masteruserrecord_test.go:156]: May I know why did you remove this subtest?

File: pkg/controller/usersignup/usersignup_controller_test.go [hunk 5]
```diff
@@ -1950,91 +1950,3 @@ func TestChangedCompliantUsername(t *testing.T) {
...
-		_, err := r.Reconcile(req)
-
-		// then verify that the MUR exists and is complete
-		require.Error(t, err)
-		murs := &v1alpha1.MasterUserRecordList{}
-		err = r.client.List(context.TODO(), murs)
-		require.NoError(t, err)
-		require.Len(t, murs.Items, 1)
-		assert.NotEqual(t, *expectedMur, murs.Items[0])
-	})
```
Comment (MatousJobanek) [inline pkg/controller/usersignup/usersignup_controller_test.go:2039]: I know that you removed the migration of `TemplateRef` part, but you could verify (also the error case) situation when either `NSLimit` or `TierName` is missing.

File: pkg/controller/masteruserrecord/masteruserrecord_controller_test.go (hunk 1 above)
Comment (tinakurian) [reply pkg/controller/masteruserrecord/masteruserrecord_controller_test.go]: fixed as part of: 000a2b5594241a958671b11904043853eb3fbb47

File: pkg/controller/masteruserrecord/sync_test.go (hunk 3 above)
Comment (tinakurian) [reply pkg/controller/masteruserrecord/sync_test.go]: fixed as part of: 000a2b5594241a958671b11904043853eb3fbb47

**Comment by tinakurian** [comment]
> As I have already asked you in the slack channel, could you please rename `migrateMurIfNecessary` to `migrateOrFixMurIfNecessary`

fixed as part of: 000a2b5594241a958671b11904043853eb3fbb47

File: pkg/controller/usersignup/masteruserrecord_test.go (hunk 4 above)
Comment (tinakurian) [reply pkg/controller/usersignup/masteruserrecord_test.go:156]: @MatousJobanek this was removed due to the removal of the migration of TemplateRef from `migrateMurIfNecessary`

Comment (tinakurian) [reply pkg/controller/usersignup/masteruserrecord_test.go]: @MatousJobanek ohhh sorry, you're right, i've re-added it here: ef347313b48735e0e6baa1c6b8bc62e28491a3f0

File: pkg/controller/masteruserrecord/sync_test.go (hunk 2 above)
Comment (tinakurian) [reply pkg/controller/masteruserrecord/sync_test.go]: good catch! added here: 3e316e28db125cd9b1d08f0e366e910b9b54fc69

File: pkg/controller/usersignup/usersignup_controller_test.go (hunk 5 above)
Comment (tinakurian) [reply pkg/controller/usersignup/usersignup_controller_test.go:2039]: @MatousJobanek i've added a test to make sure that the migration of nslimit and tiername happens here: 068aed7c2f3084cd0a0be2c71ca8837e39fd5b3a - is this good?

File: pkg/controller/masteruserrecord/sync_test.go [hunk 6]
```diff
@@ -86,14 +86,11 @@ func setupSynchronizerItems() (toolchainv1alpha1.MasterUserRecord, toolchainv1al
 	base := toolchainv1alpha1.UserAccountSpecBase{
 		NSLimit: "limit",
 		NSTemplateSet: toolchainv1alpha1.NSTemplateSetSpec{
-			TierName: "basic",
-			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{
-				Revision: "654321a",
-			},
+			TierName:         "basic",
+			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{},
```
Comment (MatousJobanek) [inline pkg/controller/masteruserrecord/sync_test.go:90]: why empty? The `ClusterResources` should be either `nil` or it should contain some `TemplateRef` set.
```suggestion
			ClusterResources: &toolchainv1alpha1.NSTemplateSetClusterResources{
				TemplateRef: "basic-clusterresources-654321a",
			},
```

File: pkg/controller/usersignup/usersignup_controller_test.go [hunk 7]
```diff
@@ -1962,54 +1962,31 @@ func TestMigrateMur(t *testing.T) {
...
-		{
-			TemplateRef: "basic-extra-123abc3",
-		},
-	}
+	expectedMur.Spec.UserAccounts[0].Spec.NSTemplateSet.TierName = "basic"
+	expectedMur.Spec.UserAccounts[0].Spec.NSLimit = "default"
+
 	expectedMur.Spec.UserAccounts[0].Spec.NSTemplateSet.ClusterResources = &toolchainv1alpha1.NSTemplateSetClusterResources{
 		TemplateRef: "basic-clusterresources-654321b",
 	}
```
Comment (MatousJobanek) [inline pkg/controller/usersignup/usersignup_controller_test.go:1982]: This should be:
```go
	// set tier name to be empty
	mur.Spec.UserAccounts[0].Spec.NSTemplateSet = toolchainv1alpha1.NSTemplateSetSpec{}
	mur.Spec.UserAccounts[0].Spec.NSLimit = ""

	expectedMur := mur.DeepCopy()
	expectedMur.Generation = 1
	expectedMur.ResourceVersion = "1"
	expectedMur.Spec.UserAccounts[0].Spec.NSTemplateSet.TierName = "basic"
	expectedMur.Spec.UserAccounts[0].Spec.NSLimit = "default"
	expectedMur.Spec.UserAccounts[0].Spec.NSTemplateSet.Namespaces = []v1alpha1.NSTemplateSetNamespace{
		{
			TemplateRef: "basic-code-123abc1",
		},
		{
			TemplateRef: "basic-dev-123abc2",
		},
		{
			TemplateRef: "basic-stage-123abc3",
		},
	}
	expectedMur.Spec.UserAccounts[0].Spec.NSTemplateSet.ClusterResources = &toolchainv1alpha1.NSTemplateSetClusterResources{
		TemplateRef: "basic-clusterresources-654321b",
	}
```

Comment (tinakurian) [reply pkg/controller/usersignup/usersignup_controller_test.go]: fixed here: d98f4bc8c02bfd5fb7180b49b944aa70a5dc37b2

File: pkg/controller/masteruserrecord/sync_test.go (hunk 6 above)
Comment (tinakurian) [reply pkg/controller/masteruserrecord/sync_test.go]: fixed here: d98f4bc8c02bfd5fb7180b49b944aa70a5dc37b2

File: go.mod [hunk 8]
```diff
@@ -64,4 +64,8 @@ replace (
 	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20200204173128-addea2498afe // avoids case-insensitive import collision: "github.com/googleapis/gnostic/openapiv2" and "github.com/googleapis/gnostic/OpenAPIv2"
 )
 
+replace github.com/codeready-toolchain/api v0.0.0-20200528065459-3e3f8c07df39 => github.com/tinakurian/api v0.0.0-20200528175500-de4844132cd6
```
Comment (xcoulon) [inline go.mod:67]: reminder ;)

Comment (tinakurian) [reply go.mod:67]: thanks @xcoulon :-)
